--- a/net/minecraft/world/entity/SpawnPlacements.java
+++ b/net/minecraft/world/entity/SpawnPlacements.java
@@ -47,6 +_,7 @@
     private static final Map<EntityType<?>, SpawnPlacements.Data> DATA_BY_TYPE = Maps.newHashMap();
 
     private static <T extends Mob> void register(
+    @Deprecated // FORGE: use SpawnPlacementRegisterEvent to register and modify spawn placements
         EntityType<T> entityType, SpawnPlacementType spawnPlacementType, Heightmap.Types heightmapType, SpawnPlacements.SpawnPredicate<T> predicate
     ) {
         SpawnPlacements.Data spawnplacements$data = DATA_BY_TYPE.put(entityType, new SpawnPlacements.Data(heightmapType, spawnPlacementType, predicate));
@@ -73,7 +_,8 @@
         EntityType<T> entityType, ServerLevelAccessor serverLevel, MobSpawnType spawnType, BlockPos pos, RandomSource random
     ) {
         SpawnPlacements.Data spawnplacements$data = DATA_BY_TYPE.get(entityType);
-        return spawnplacements$data == null || ((SpawnPredicate<T>)spawnplacements$data.predicate).test(entityType, serverLevel, spawnType, pos, random);
+        boolean vanillaResult = spawnplacements$data == null || spawnplacements$data.predicate.test((EntityType)entityType, serverLevel, spawnType, pos, random);
+        return net.neoforged.neoforge.event.EventHooks.checkSpawnPlacements(entityType, serverLevel, spawnType, pos, random, vanillaResult);
     }
 
     static {
@@ -168,5 +_,14 @@
     @FunctionalInterface
     public interface SpawnPredicate<T extends Entity> {
         boolean test(EntityType<T> entityType, ServerLevelAccessor serverLevel, MobSpawnType spawnType, BlockPos pos, RandomSource random);
+    }
+
+    // ******* FORGE START. INTERNAL USE ONLY! ****** //
+    public static void fireSpawnPlacementEvent()
+    {
+         Map<EntityType<?>, net.neoforged.neoforge.event.entity.SpawnPlacementRegisterEvent.MergedSpawnPredicate<?>> map = Maps.newHashMap();
+         DATA_BY_TYPE.forEach((type, data) -> map.put(type, new net.neoforged.neoforge.event.entity.SpawnPlacementRegisterEvent.MergedSpawnPredicate<>(data.predicate, data.placement, data.heightMap)));
+         net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.event.entity.SpawnPlacementRegisterEvent(map));
+         map.forEach(((entityType, merged) -> DATA_BY_TYPE.put(entityType, new SpawnPlacements.Data(merged.getHeightmapType(), merged.getSpawnType(), merged.build()))));
     }
 }
