--- a/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
+++ b/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
@@ -106,6 +_,7 @@
     }
 
     public boolean destroyBlock(BlockPos p_105268_) {
+        if (minecraft.player.getMainHandItem().onBlockStartBreak(p_105268_, minecraft.player)) return false;
         if (this.minecraft.player.blockActionRestricted(this.minecraft.level, p_105268_, this.localPlayerMode)) {
             return false;
         } else {
@@ -120,9 +_,8 @@
                 } else if (blockstate.isAir()) {
                     return false;
                 } else {
-                    block.playerWillDestroy(level, p_105268_, blockstate, this.minecraft.player);
                     FluidState fluidstate = level.getFluidState(p_105268_);
-                    boolean flag = level.setBlock(p_105268_, fluidstate.createLegacyBlock(), 11);
+                    boolean flag = blockstate.onDestroyedByPlayer(level, p_105268_, minecraft.player, false, fluidstate);
                     if (flag) {
                         block.destroy(level, p_105268_, blockstate);
                     }
@@ -270,7 +_,7 @@
 
     private boolean sameDestroyTarget(BlockPos p_105282_) {
         ItemStack itemstack = this.minecraft.player.getMainHandItem();
-        return p_105282_.equals(this.destroyBlockPos) && ItemStack.isSameItemSameComponents(itemstack, this.destroyingItem);
+        return p_105282_.equals(this.destroyBlockPos) && !destroyingItem.shouldCauseBlockBreakReset(itemstack);
     }
 
     private void ensureHasSentCarriedItem() {
