--- a/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
+++ b/net/minecraft/client/multiplayer/MultiPlayerGameMode.java
@@ -106,6 +_,7 @@
     }
 
     public boolean destroyBlock(BlockPos p_105268_) {
+        if (minecraft.player.getMainHandItem().onBlockStartBreak(p_105268_, minecraft.player)) return false;
         if (this.minecraft.player.blockActionRestricted(this.minecraft.level, p_105268_, this.localPlayerMode)) {
             return false;
         } else {
@@ -120,9 +_,8 @@
                 } else if (blockstate.isAir()) {
                     return false;
                 } else {
-                    block.playerWillDestroy(level, p_105268_, blockstate, this.minecraft.player);
                     FluidState fluidstate = level.getFluidState(p_105268_);
-                    boolean flag = level.setBlock(p_105268_, fluidstate.createLegacyBlock(), 11);
+                    boolean flag = blockstate.onDestroyedByPlayer(level, p_105268_, minecraft.player, false, fluidstate);
                     if (flag) {
                         block.destroy(level, p_105268_, blockstate);
                     }
@@ -270,7 +_,7 @@
 
     private boolean sameDestroyTarget(BlockPos p_105282_) {
         ItemStack itemstack = this.minecraft.player.getMainHandItem();
-        return p_105282_.equals(this.destroyBlockPos) && ItemStack.isSameItemSameComponents(itemstack, this.destroyingItem);
+        return p_105282_.equals(this.destroyBlockPos) && !destroyingItem.shouldCauseBlockBreakReset(itemstack);
     }
 
     private void ensureHasSentCarriedItem() {
@@ -402,7 +_,10 @@
         this.ensureHasSentCarriedItem();
         Vec3 vec3 = p_105233_.getLocation().subtract(p_105232_.getX(), p_105232_.getY(), p_105232_.getZ());
         this.connection.send(ServerboundInteractPacket.createInteractionPacket(p_105232_, p_105231_.isShiftKeyDown(), p_105234_, vec3));
-        return this.localPlayerMode == GameType.SPECTATOR ? InteractionResult.PASS : p_105232_.interactAt(p_105231_, vec3, p_105234_);
+        if (this.localPlayerMode == GameType.SPECTATOR) return InteractionResult.PASS;
+        var fromEvent = net.neoforged.neoforge.event.EventHooks.onEntityInteract(p_105231_, p_105231_, p_105234_, vec3);
+        if (fromEvent != null) return fromEvent;
+        return p_105232_.interactAt(p_105231_, vec3, p_105234_);
     }
 
     public void handleInventoryMouseClick(int p_171800_, int p_171801_, int p_171802_, ClickType p_171803_, Player p_171804_) {
