--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -37,7 +_,7 @@
         Level level = context.getLevel();
         BlockPos blockpos = context.getClickedPos();
         BlockPos blockpos1 = blockpos.relative(context.getClickedFace());
-        if (growCrop(context.getItemInHand(), level, blockpos)) {
+        if (applyBonemeal(context.getItemInHand(), level, blockpos, context.getPlayer())) {
             if (!level.isClientSide) {
                 context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH);
                 level.levelEvent(1505, blockpos, 15);
@@ -60,15 +_,24 @@
         }
     }
 
+    @Deprecated //Forge: Use Player/Hand version
     public static boolean growCrop(ItemStack stack, Level level, BlockPos pos) {
-        BlockState blockstate = level.getBlockState(pos);
-        if (blockstate.getBlock() instanceof BonemealableBlock bonemealableblock && bonemealableblock.isValidBonemealTarget(level, pos, blockstate)) {
-            if (level instanceof ServerLevel) {
-                if (bonemealableblock.isBonemealSuccess(level, level.random, pos, blockstate)) {
-                    bonemealableblock.performBonemeal((ServerLevel)level, level.random, pos, blockstate);
+        if (level instanceof net.minecraft.server.level.ServerLevel)
+            return applyBonemeal(stack, level, pos, null);
+        return false;
+    }
+
+    public static boolean applyBonemeal(ItemStack p_40628_, Level p_40629_, BlockPos p_40630_, @Nullable net.minecraft.world.entity.player.Player player) {
+        BlockState blockstate = p_40629_.getBlockState(p_40630_);
+        var event = net.neoforged.neoforge.event.EventHooks.fireBonemealEvent(player, p_40629_, p_40630_, blockstate, p_40628_);
+        if (event.isCanceled()) return event.isSuccessful();
+        if (blockstate.getBlock() instanceof BonemealableBlock bonemealableblock && bonemealableblock.isValidBonemealTarget(p_40629_, p_40630_, blockstate)) {
+            if (p_40629_ instanceof ServerLevel) {
+                if (bonemealableblock.isBonemealSuccess(p_40629_, p_40629_.random, p_40630_, blockstate)) {
+                    bonemealableblock.performBonemeal((ServerLevel)p_40629_, p_40629_.random, p_40630_, blockstate);
                 }
 
-                stack.shrink(1);
+                p_40628_.shrink(1);
             }
 
             return true;
