--- a/net/minecraft/world/item/ShieldItem.java
+++ b/net/minecraft/world/item/ShieldItem.java
@@ -54,6 +_,13 @@
         return repair.is(ItemTags.PLANKS) || super.isValidRepairItem(toRepair, repair);
     }
 
+    /* ******************** FORGE START ******************** */
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_SHIELD_ACTIONS.contains(itemAbility);
+    }
+
     @Override
     public EquipmentSlot getEquipmentSlot() {
         return EquipmentSlot.OFFHAND;
