--- a/net/minecraft/recipebook/PlaceRecipe.java
+++ b/net/minecraft/recipebook/PlaceRecipe.java
@@ -9,7 +_,7 @@
     default void placeRecipe(int width, int height, int outputSlot, RecipeHolder<?> recipe, Iterator<T> ingredients, int maxAmount) {
         int i = width;
         int j = height;
-        if (recipe.value() instanceof ShapedRecipe shapedrecipe) {
+        if (recipe.value() instanceof net.neoforged.neoforge.common.crafting.IShapedRecipe<?> shapedrecipe) {
             i = shapedrecipe.getWidth();
             j = shapedrecipe.getHeight();
         }
